### DISCUSSION TITLE
Nue 1.0 (Beta)

### DISCUSSION THREAD

**Discussion by tipiirai** [nuejs/nue#318] [Announcements]
A new, *massive* release is now complete. Please check out the release notes: https://nuejs.org/blog/nue-1-beta/

<a href="https://nuejs.org/blog/nue-1-beta/">
  <img src="https://nuejs.org/img/og-blue-big.png" width="750"></a><br><br>

Would love to hear your thoughts. Thank you!

**Comment by seneca** [comment]
This is great news but I'm getting errors:

xxxx Nue % nue create simple-blog

Loading template...
✓ Initialize simple-blog/.dist/dev: •••••ENOENT: No such file or directory
   errno: -2
 syscall: "stat"
   path: "/xxx/xxxx/.bun/install/global/node_modules/nue-glow/minified/syntax.css"

**Reply by tipiirai** [reply]
Do you have the latest 1.0.0-beta? Not sure, but upgrading should work like this:

```
bun i --global nuekit@latest
```

Check with `nue --version`

**Comment by seneca** [comment]
And when I want to spin the server:

xxx simple-blog % nue

✓ Nue 1.0.0-beta • Bun 1.1.24 
✓ Serving site from .dist/dev
✓ http://localhost:8083/ 
✓ /@global/colors.css 
✓ /@global/layout.css 
✓ /@global/typography.css 
✓ /@library/gallery.css 
✓ /@library/motion.css 
Error on <header> component 169 | function processChild(comp, node, deps, data) {
170 |   const { attribs } = node
171 | 
172 |   // merge attributes
173 |   const child = comp.create({ ...data, ...attribs }, deps, node.children)
174 |   if (child.children.length == 1) mergeAttribs(child.firstChild.attribs, attribs)
                  ^
TypeError: undefined is not an object (evaluating 'child.children.length')
      at processChild (/xxx/xxx/.bun/install/global/node_modules/nuejs-core/src/render.js:174:13)
      at walk (//xxx/xxx/.bun/install/global/node_modules/nuejs-core/src/render.js:270:22)
      at walkChildren (/xxx/xxx/.bun/install/global/node_modules/nuejs-core/src/render.js:278:15)
      at walk (/xxx/xxx/.bun/install/global/node_modules/nuejs-core/src/render.js:203:7)
      at processNode (/xxx/xxx/.bun/install/global/node_modules/nuejs-core/src/render.js:282:3)
      at /xxx/xxx/.bun/install/global/node_modules/nuejs-core/src/render.js:162:23
      at forEach (1:11)
      at processFor (/xxx/xxx/.bun/install/global/node_modules/nuejs-core/src/render.js:142:9)
      at walkChildren (/xxx/xxx/.bun/install/global/node_modules/nuejs-core/src/render.js:278:15)
      at walk (/xxx/xxx/.bun/install/global/node_modules/nuejs-core/src/render.js:231:7)

{
  component: "header",
  originalLine: 110,
  originalColumn: 21,
}
Not found /favicon.ico
Not found /apple-touch-icon-precomposed.png
Not found /apple-touch-icon.png

**Comment by seneca** [comment]
I've done:
xxx@xxx-iMac ~ % bun i --global nuekit@latest
bun add v1.1.24-canary.14 (b70458c3)

installed nuekit@1.0.0-beta with binaries:
 - nue
 [632.00ms] done


and then
xxx@xxx-iMac Nue % nue create simple-blog
Loading template...
✓ Initialize simple-blog/.dist/dev: •••••ENOENT: No such file or directory
   errno: -2
 syscall: "stat"
   path: "/xxx/xxx/.bun/install/global/node_modules/nue-glow/minified/syntax.css"

// Will try to poke why this is happening later.

**Comment by sloul1** [comment]
I had also some issues after updating and creating “simple-blog”. Tested on couple years old AMD Ryzen 7 (5700U) based Lenovo laptop Ubuntu 22.04 LTS workstation as os. Tested also on newly created and updated VM Ubuntu 24.04 LTS workstation on Proxmox which worked without hitch. I’ve been testing earlier version on my laptop and suspected problem to be somehow residual of earlier tests. At some point I used command ```shell bun i --global nuekit@latest ``` provided in this conversation. On laptop I got new simple-blog working after deleting .bun directory and installing bun and nue using three command combination from current documentation. I did some testing also on Android tablet with Termux but I won’t get into details on this at this moment.
 ```shell
~$ sudo rm -rf /home/user/.bun
curl -fsSL https://bun.sh/install | bash   
bun install nuekit --global
nue create simple-blog
```
Ubuntu 22.04 testing below.
```shell
$ uname -a
Linux xxx-xxx 6.5.0-45-generic #45~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Mon Jul 15 16:40:02 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```
```shell
~/nuete/simple-blog$ nue

✓ Nue 1.0.0-beta • Bun 1.1.24 
✓ Serving site from .dist/dev
✓ http://localhost:8083/ 
✓ /@global/colors.css 
✓ /@global/layout.css 
✓ /@global/typography.css 
✓ /@library/gallery.css 
✓ /@library/motion.css 
Error on <header> component 169 | function processChild(comp, node, deps, data) {
170 |   const { attribs } = node
171 | 
172 |   // merge attributes
173 |   const child = comp.create({ ...data, ...attribs }, deps, node.children)
174 |   if (child.children.length == 1) mergeAttribs(child.firstChild.attribs, attribs)
                  ^
TypeError: undefined is not an object (evaluating 'child.children.length')
      at processChild (/home/pepper/.bun/install/global/node_modules/nuejs-core/src/render.js:174:13)
      at walk (/home/pepper/.bun/install/global/node_modules/nuejs-core/src/render.js:270:22)
      at walkChildren (/home/pepper/.bun/install/global/node_modules/nuejs-core/src/render.js:278:15)
      at walk (/home/pepper/.bun/install/global/node_modules/nuejs-core/src/render.js:203:7)
      at processNode (/home/pepper/.bun/install/global/node_modules/nuejs-core/src/render.js:282:3)
      at /home/pepper/.bun/install/global/node_modules/nuejs-core/src/render.js:162:23
      at forEach (1:11)
      at processFor (/home/pepper/.bun/install/global/node_modules/nuejs-core/src/render.js:142:9)
      at walkChildren (/home/pepper/.bun/install/global/node_modules/nuejs-core/src/render.js:278:15)
      at walk (/home/pepper/.bun/install/global/node_modules/nuejs-core/src/render.js:231:7)

{
  component: "header",
  originalLine: 110,
  originalColumn: 21,
}
^C
```
NEW TRY
```shell
~/nuete/simple-blog$ cd ..
~/nuete$ rm -rf simple-blog/
~/nuete$ nue create simple-blog

Loading template...
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.provenance'
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.provenance'
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.provenance'
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.lastuseddate#PS'
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.lastuseddate#PS'
tar: Ignoring unknown extended header keyword 'LIBARCHIVE.xattr.com.apple.provenance'
✓ Initialize simple-blog/.dist/dev: •••••ENOENT: No such file or directory
   errno: -2
 syscall: "open"
```

**Reply by tipiirai** [reply]
Thank you for the details! Will look into this tomorrow.

**Reply by nobkd** [reply]
I think the first error is because we have the versions of `nue...` on `*` instead of `latest`, and it just reuses currently installed versions. (result: out of sync APIs)
We changed that to `*` for testing purposes, because `latest` uses the version from npm registry instead of the current workspace...I at least think, that was the reason.
Maybe just a script to replace `*` with `latest` on release?

---

>  I did some testing also on Android tablet with Termux but I won’t get into details on this at this moment.

I think termux won't work, because there is no lightningcss version compiled for it (at least with node). Will check it again. Maybe with bun it can work, but I haven't got bun working on termux, so :man_shrugging:

---

I might update this comment...

**Reply by tipiirai** [reply]
So while I run `npm publish` locally the package.json files should have "latest", but on git, we should have "*"?

**Reply by nobkd** [reply]
Yes. I think so.

---

PR where changed to `*`
- #261

Comment (by me) mentioning, it could cause problems on release :sweat_smile:
- https://github.com/nuejs/nue/pull/261#issuecomment-2058946728

Comment (by me, too) disregarding the release issues :sweat_smile: :sweat_smile: xD
- https://github.com/nuejs/nue/pull/261#issuecomment-2060778502 ()

Now feels like a prophecy:
> but I'm not sure, if it is that great, because **cached versions** could also be used with `*` instead of real `latest`

---

Here's a little (bad) script to run before publishing (maybe through `prepublishOnly`-`package.json` script)... (to make the inconvenience more convenient)

<details>
<summary>script</summary>

(run from root directory)

```js
import { readdir, readFile, writeFile } from 'node:fs/promises'
import { join } from 'node:path'

const dir = 'packages'
const pkgs = (await readdir(dir, { recursive: true }))
    .filter(pth => pth.endsWith('package.json'))
    .map(pth => join(dir, pth))

for (const pkg of pkgs) {
    const json = JSON.parse(await readFile(pkg, { encoding: 'utf8' }))
    Object.keys(json.dependencies ?? {})
        .filter(key => key.startsWith('nue'))
        .forEach(key => json.dependencies[key] = 'latest')
    await writeFile(pkg, JSON.stringify(json, null, 2) + '\n')
}
```

</details>

**Reply by tipiirai** [reply]
Gotcha. One thing that I've been thinking about recently is to get rid of npm packages and the monorepo setup and simply place all the code into one repository. That is: glow, nuemark, and nuejs are bundled inside a single project. There would be one package.json. Of course, the projects could not be used in isolation then, but that would cause fewer problems and simplify the publishing process. What do you think?

**Reply by nobkd** [reply]
Actually I quite like that the main features are split in packages. But I also understand that it can be a maintenance burden. E.g. because of */latest

At least glow should stay it's own package, because there are already some people using the highlighter to totally unrelated things from nue because it is just super small and works great through its relatively simple pattern matching style. (See e.g. [here](https://www.npmjs.com/package/glow-highlighter)) 

(Maybe we can use [bundled dependencies](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#bundledependencies) for nuekit releases, so it already contains all nue* stuff needed? Not sure if that makes sense... Never done that...)

(Also I think this discussion shouldn't be done in this discussion thread ;)

In the end, it's still your decision, and what you like best.

(ps: will probably be away for a few days **after** tomorrow)

**Reply by tipiirai** [reply]
I just published `nuekit-1.0.0-beta.1` and `nuemark-0.4.1` using "latest" and brought them back to "*" after pushing to github. It was actually pretty trivial to do manually, just need to remember the next time. So I'll keep using this workflow.  `bundleDependencies` sounds good, but I have my doubts if Bun install supports them. 

Thanks for the help here! Pretty crucial stuff. I'm also away next week.

**Reply by nobkd** [reply]
@sloul1 can you check if it works with the new beta release? Thank you.

**Reply by sloul1** [reply]
> @sloul1 can you check if it works with the new beta release? Thank you.

I did some testing today. I'll get back to this rested, clear mind. I'd say tomorrow or during this weekend.

**Reply by sloul1** [reply]
> @sloul1 can you check if it works with the new beta release? Thank you.

I just tested again and following worked without problems on both Ubuntu 22.04 LTS laptop and Ubuntu 24.04 LTS vm I mentioned earlier. =)
```
$ sudo rm -rf ~/.bun
$ bun install nuekit --global
bun add v1.1.24 (85a32991)

installed nuekit@1.0.0-beta.1 with binaries:
 - nue

20 packages installed [4.44s]
```
```
$ nue -v
✓ Nue 1.0.0-beta.1 • Bun 1.1.24
$ nue create simple-blog
```
Thank you.

**Comment by seneca** [comment]
I've just run the latest beta update and it works great. Thanks.
